src/Rings/groebner.jl

```diff
@@ -1221,3 +1231,112 @@ function _compute_groebner_basis_using_fglm(I::MPolyIdeal,
 	dim(I) == 0 || error("Dimension of ideal must be zero.")
 	I.gb[destination_ordering] = _fglm(G, destination_ordering)
 end
+
+
+@doc Markdown.doc"""
+    groebner_basis_hilbert_driven(I::MPolyIdeal,
+                                  destination_ordering::MonomialOrdering;
+                                  weights::Vector{E} = ones(Int, ngens(base_ring(I))),
+                                  complete_reduction::Bool = false) where {E <: Integer}
+
+
+Compute a Gröbner basis of `I` with respect to `destination_ordering`
+using a Hilbert Series driven method as follows: If a Gröbner basis
+for `I` is present, compute the Hilbert series of `I` and use it to
+optimize the Gröbner basis computation for `I`
+w.r.t. `destination_ordering`. If no Gröbner basis for `I` is present
+compute the Hilbert series for `I` if the base field of `I` has
+positive characteristic, otherwise compute the Hilbert series for `I`
+modulo a randomly chosen prime. Use the resulting Hilbert series to
+optimize the Gröbner basis computation for `I`
+w.r.t. `target_ordering`.
+
+`I` must be given by generators homogeneous w.r.t. `weights`.
+
+# Examples
+```jldoctest
+julia> R, (x, y, z) = PolynomialRing(QQ, ["x", "y", "z"]);
+
+julia> I = ideal(R, [x^2 + y*z, x*y - y*z]);
+
+julia> groebner_basis_hilbert_driven(I, lex(R), complete_reduction = true)
+Gröbner basis with elements
+1 -> y^2*z + y*z^2
+2 -> x*y - y*z
+3 -> x^2 + y*z
+with respect to the ordering
+lex([x, y, z])
+```
+"""
+
+function groebner_basis_hilbert_driven(I::MPolyIdeal,
+                                       destination_ordering::MonomialOrdering;
+                                       weights::Vector{E} = ones(Int, ngens(base_ring(I))),
+                                       complete_reduction::Bool = false) where {E <: Integer}
+  
+  all(p -> _is_homogeneous_weights(p, weights), gens(I)) || error("I must be given by generators homogeneous with respect to given weights.")
+	isa(coefficient_ring(base_ring(I)), AbstractAlgebra.Field) || error("The underlying coefficient ring of I must be a field.")
+  is_global(destination_ordering)) || error("Destination ordering must be global.")
+  typeof(destination_ordering) <: MatrixOrdering && error("Hilbert driven Gröbner basis computation not implemented for general matrix orderings.") 
+	haskey(I.gb, destination_ordering) && return I.gb[destination_ordering]
+  if isempty(I.gb) && iszero(characteristic(base_ring(I)))  
+    while true
+      r = rand(2^15:2^16)
+      p = Hecke.next_prime(r)
+        
+      base_field = GF(p)
+      ModP, _ = PolynomialRing(base_field, "x" => 1:ngens(base_ring(I)))
+      I_mod_p_gens = Vector{elem_type(ModP)}(undef, length(gens(I))) 
+      try
+        I_mod_p_gens = [ModP(f) for f in gens(I)]
+        G = f4(ideal(ModP, I_mod_p_gens))
+        break
+      catch
```

Comment:
What is the possible error? The error should be checked and rethrown if it does not fit.

Reply:
This error is caused if one of the coefficients of the input polynomials is divisible by our prime `p`. I changed the code to handle specifically only this error (see https://github.com/oscar-system/Oscar.jl/pull/1821/commits/89e6371abc0aabc6da63938f43b2f637fc9c3df6), I don't know if there is a better way to handle this.

Reply:
Looks good, thanks.